snooty/flutter.py

```diff
@@ -0,0 +1,250 @@
+        origin = getattr(ty, '__origin__', None)
+        if origin is not None:
+            args = getattr(ty, '__args__')
+            if origin is list:
+                return f'list{plural_suffix} of {inner(args[0], True, level)}'
+            elif origin is dict:
+                key_type = inner(args[0], True, level)
+                value_type = inner(args[1], True, level)
+                return f'mapping{plural_suffix} of {key_type} to {value_type}'
+            elif origin is tuple:
+                # Tuples are a hard problem... this is okay
+                return pluralize(_get_typename(ty))
+            elif origin is Union:
+                if len(args) == 2:
+                    try:
+                        none_index = args.index(type(None))
+                    except ValueError:
+                        pass
+                    else:
+                        non_none_arg = args[int(not none_index)]
+                        return f'optional {inner(non_none_arg, plural, level)}'
+
+                up_to_last = args[:-1]
+                part1 = (inner(arg, plural, level=level) for arg in up_to_last)
+                part2 = inner(args[-1], plural, level=level)
+                if not plural:
+                    part1 = (_add_indefinite_article(desc) for desc in part1)
+                    part2 = _add_indefinite_article(part2)
+                comma = ',' if len(up_to_last) > 1 else ''
+                joined_part1 = ', '.join(part1)
+                return f'either {joined_part1}{comma} or {part2}'
+
+        # A custom type
+        if ty not in hints:
+            hints[ty] = _generate_hint(ty, lambda ty: inner(ty, False, level))
+
+        try:
+            return pluralize(ty.__name__)
+        except AttributeError:
+            return pluralize(_get_typename(ty))
+
+    return inner(ty, False, 0), hints
+
+
+def checked(klass: Type[T]) -> Type[T]:
+    """Marks a dataclass as being deserializable."""
+    CACHED_TYPES[klass] = None
+    return klass
+
+
+class LoadError(TypeError):
+    def __init__(self, message: str, ty: type, bad_data: object) -> None:
+        super().__init__(message)
+        self.expected_type = ty
+        self.bad_data = bad_data
+
+
+class LoadWrongType(LoadError):
+    def __init__(self, ty: type, bad_data: object) -> None:
+        description, hints = english_description_of_type(ty)
+        hint_text = '\n\n'.join(hints.values())
+        if hint_text:
+            hint_text = '\n\n' + hint_text
+
+        super().__init__(
+            f'Incorrect type. Expected {description}.{hint_text}',
+            ty,
+            bad_data)
+
+
+class LoadWrongArity(LoadWrongType):
+    pass
+
+
+class LoadUnknownField(LoadError):
+    def __init__(self, ty: type, bad_data: object, bad_field: str) -> None:
+        super().__init__('Unexpected field: "{}"'.format(bad_field), ty, bad_data)
+        self.bad_field = bad_field
+
+
+def check_type(ty: Type[C], data: object, ty_module: str = '') -> C:
```

Comment:
What is ty_module for?  It looks like it's only passed recursively to check_type, and never used otherwise.

Reply:
Oh, interesting. I forgot why I added `ty_module`; I guess it's cruft from a past implementation